client/src/api/block_builder/input_selection/new/mod.rs

```diff
@@ -0,0 +1,297 @@
+
+    /// Creates a new [`InputSelection`].
+    pub fn new(
+        available_inputs: Vec<InputSigningData>,
+        outputs: Vec<Output>,
+        protocol_parameters: ProtocolParameters,
+    ) -> Self {
+        Self {
+            available_inputs,
+            required_inputs: None,
+            forbidden_inputs: HashSet::new(),
+            selected_inputs: Vec::new(),
+            outputs: outputs
+                .into_iter()
+                .map(|output| OutputInfo { output, provided: true })
+                .collect(),
+            burn: None,
+            remainder_address: None,
+            protocol_parameters,
+            timestamp: instant::SystemTime::now()
+                .duration_since(instant::SystemTime::UNIX_EPOCH)
+                .expect("time went backwards")
+                .as_secs() as u32,
+        }
+    }
+
+    /// Sets the required inputs of an [`InputSelection`].
+    pub fn required_inputs(mut self, inputs: HashSet<OutputId>) -> Self {
+        self.required_inputs.replace(inputs);
+        self
+    }
+
+    /// Sets the forbidden inputs of an [`InputSelection`].
+    pub fn forbidden_inputs(mut self, inputs: HashSet<OutputId>) -> Self {
+        self.forbidden_inputs = inputs;
+        self
+    }
+
+    /// Sets the burn of an [`InputSelection`].
+    pub fn burn(mut self, burn: Burn) -> Self {
+        self.burn.replace(burn);
+        self
+    }
+
+    /// Sets the remainder address of an [`InputSelection`].
+    pub fn remainder_address(mut self, address: Address) -> Self {
+        self.remainder_address.replace(address);
+        self
+    }
+
+    /// Sets the timestamp of an [`InputSelection`].
+    pub fn timestamp(mut self, timestamp: u32) -> Self {
+        self.timestamp = timestamp;
+        self
+    }
+
+    // TODO should we somehow enforce using filter so we don't have to use can_be_unlocked_now later everywhere ?
+    /// Filters out the available inputs that
+    /// - can't be unlocked by the given addresses
+    /// - can't be unlocked now
+    pub fn filter(self, addresses: &[Address]) -> Self {
+        let _addresses = addresses
+            .iter()
+            // TODO meh
+            .copied()
+            .chain(self.available_inputs.iter().filter_map(|input| match &input.output {
+                Output::Alias(output) => Some(Address::Alias(AliasAddress::from(
+                    output.alias_id_non_null(input.output_id()),
+                ))),
+                Output::Nft(output) => Some(Address::Nft(NftAddress::from(
+                    output.nft_id_non_null(input.output_id()),
+                ))),
+                _ => None,
+            }));
+
+        // self.available_inputs.retain(|input| input.can_be_unlocked_now(input, addresses, ???, self.time))
+
+        self
+    }
+
+    /// Selects inputs that meet the requirements of the outputs to satisfy the semantic validation of the overall
+    /// transaction. Also creates outputs if transitions are required.
```

Comment:
Maybe something like this to include the remainder 
```suggestion
    /// transaction. Also creates a remainder output and chain transition outputs if required.
```